src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php

```diff
@@ -41,39 +41,82 @@ trait PriorityTaggedServiceTrait
      */
     private function findAndSortTaggedServices($tagName, ContainerBuilder $container)
     {
-        $indexAttribute = $defaultIndexMethod = $needsIndexes = null;
+        $indexAttribute = $defaultIndexMethod = $needsIndexes = $defaultPriorityMethod = null;
 
         if ($tagName instanceof TaggedIteratorArgument) {
             $indexAttribute = $tagName->getIndexAttribute();
             $defaultIndexMethod = $tagName->getDefaultIndexMethod();
             $needsIndexes = $tagName->needsIndexes();
+            $defaultPriorityMethod = $tagName->getDefaultPriorityMethod();
             $tagName = $tagName->getTag();
         }
 
         $services = [];
 
         foreach ($container->findTaggedServiceIds($tagName, true) as $serviceId => $attributes) {
-            $priority = isset($attributes[0]['priority']) ? $attributes[0]['priority'] : 0;
+            $class = $container->getDefinition($serviceId)->getClass();
+            $class = $container->getParameterBag()->resolveValue($class) ?: null;
+
+            if (!$r = $container->getReflectionClass($class)) {
+                throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $serviceId));
+            }
+
+            $priority = 0;
+            if ($defaultPriorityMethod && $r->hasMethod($defaultPriorityMethod)) {
+                if (!($rpm = $r->getMethod($defaultPriorityMethod))->isStatic()) {
+                    throw new InvalidArgumentException(
+                        sprintf(
+                            'Method "%s::%s()" should be static: tag "%s" on service "%s" is missing default priority method.',
```

Comment:
I think that `tag ... is missing` is not actually correct. It's not **missing** it, since the method does exist, it's just not accessible. What if we try smth like this: 
```suggestion
                            'Default priority method "%s::%s()" of the "%s"-tagged collection on service "%s" must be static.',
```